src/wallet/sqlite.cpp

```diff
@@ -58,44 +52,96 @@ static int TraceSqlCallback(unsigned code, void* context, void* param1, void* pa
     return SQLITE_OK;
 }
 
-static bool BindBlobToStatement(sqlite3_stmt* stmt,
-                                int index,
-                                std::span<const std::byte> blob,
-                                const std::string& description)
+template <typename T>
+concept ColumnBlob = requires(T a, T::element_type* data, size_t size)
 {
-    // Pass a pointer to the empty string "" below instead of passing the
-    // blob.data() pointer if the blob.data() pointer is null. Passing a null
-    // data pointer to bind_blob would cause sqlite to bind the SQL NULL value
-    // instead of the empty blob value X'', which would mess up SQL comparisons.
-    int res = sqlite3_bind_blob(stmt, index, blob.data() ? static_cast<const void*>(blob.data()) : "", blob.size(), SQLITE_STATIC);
-    if (res != SQLITE_OK) {
-        LogPrintf("Unable to bind %s to statement: %s\n", description, sqlite3_errstr(res));
-        sqlite3_clear_bindings(stmt);
-        sqlite3_reset(stmt);
-        return false;
+    T(data, size);
+};
+
+/** RAII class that encapsulates a sqlite3_stmt */
+class SQLiteStatement
+{
+private:
+    sqlite3* m_db{nullptr};
+    sqlite3_stmt* m_stmt{nullptr};
+
+public:
+    explicit SQLiteStatement(sqlite3* db, const std::string& stmt_text)
+        : m_db(db)
+    {
+        int res = sqlite3_prepare_v2(m_db, stmt_text.c_str(), -1, &m_stmt, nullptr);
+        if (res != SQLITE_OK) {
+            throw std::runtime_error(strprintf(
+                "SQLiteDatabase: Failed to prepare SQL statement: %s\n", sqlite3_errstr(res)));
+        }
     }
 
-    return true;
-}
+    ~SQLiteStatement()
+    {
+        Reset();
+        sqlite3_finalize(m_stmt);
+    }
+
+    int Step()
+    {
+        return sqlite3_step(m_stmt);
+    }
+
+    void Reset()
+    {
+        sqlite3_clear_bindings(m_stmt);
+        sqlite3_reset(m_stmt);
+    }
+
+    bool Bind(int index, std::span<const std::byte> data, const std::string& description)
+    {
+        // Pass a pointer to the empty string "" below instead of passing the
+        // blob.data() pointer if the blob.data() pointer is null. Passing a null
+        // data pointer to bind_blob would cause sqlite to bind the SQL NULL value
+        // instead of the empty blob value X'', which would mess up SQL comparisons.
+        int res = sqlite3_bind_blob(m_stmt, index, data.data() ? static_cast<const void*>(data.data()) : "", data.size(), SQLITE_STATIC);
+        if (res != SQLITE_OK) {
+            LogPrintf("Unable to bind %s to statement: %s\n", description, sqlite3_errstr(res));
+            Reset();
+            return false;
+        }
+
+        return true;
+    }
+
+    template<typename T>
+    T Column(int col)
+    {
+        if constexpr (std::integral<T> && sizeof(T) <= 4) {
+            return sqlite3_column_int(m_stmt, col);
+        } else if constexpr (std::integral<T> && sizeof(T) <= 8) {
+            return static_cast<int64_t>(sqlite3_column_int64(m_stmt, col));
```

Comment:
What happens if `T` is unsigned? Maybe add https://en.cppreference.com/w/cpp/types/is_signed.html here: `std::integral<T> && std::is_signed<T>`? (and leave it to drop to "else assert always false" for unsigned `T`s)

Reply:
Done